core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -498,12 +499,23 @@ class ProducerStateManager(val topicPartition: TopicPartition,
   private var lastMapOffset = 0L
   private var lastSnapOffset = 0L
 
+  // Keep track of the last timestamp from the oldest transaction. This is used
+  // to detect (approximately) when a transaction has been left hanging on a partition.
+  // We make the field volatile so that it can be safely accessed without a lock.
+  @volatile private var oldestTxnLastTimestamp: Long = -1L
+
   // ongoing transactions sorted by the first offset of the transaction
   private val ongoingTxns = new util.TreeMap[Long, TxnMetadata]
 
   // completed transactions whose markers are at offsets above the high watermark
   private val unreplicatedTxns = new util.TreeMap[Long, TxnMetadata]
 
+  @threadsafe
+  def hasLateTransaction(currentTimeMs: Long): Boolean = {
+    val lastTimestamp = oldestTxnLastTimestamp
+    lastTimestamp > 0 && (currentTimeMs - lastTimestamp) > maxTransactionTimeoutMs
```

Comment:
I have a meta question. In the KIP, you wrote: "This metric will record the number of partitions which have open transactions with durations exceeding `transaction.max.timeout.ms` (plus 5 minutes)." It seems that the "+5 minutes" is not applied anywhere. I guess that proposed this to avoid counting transactions that are about to be expired or are being expired. Have you changed your mind on this point?

Reply:
Yeah, good question. I have been debating it. I had to rely on the last timestamp here instead of the transaction start time, so that already adds a slop factor. But is it enough? Perhaps it is better to still add the 5 minutes to avoid the likelihood of false positives. What do you think?

Reply:
I would add the 5 minutes to avoid the false positives but I don't feel super strong about it.

Reply:
@dajac Thanks, I pushed a commit to add the 5s buffer. It is what the KIP specifies after all and I couldn't come up with a strong reason not to have it.